File: csharp/binary-search-tree/BinarySearchTree.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

public class BinarySearchTree : IEnumerable<int>
{
    public BinarySearchTree(int value) => Value = value;

    public BinarySearchTree(IEnumerable<int> values) : this(values.First())
    {
        foreach(var value in values.Skip(1))
            Add(value);
    }

    public BinarySearchTree Add(int value)
    {
        if(value <= Value)
            Left = Left != null ? Left.Add(value) : new BinarySearchTree(value);
        else
            Right = Right != null ? Right.Add(value) : new BinarySearchTree(value);
        return this;
    }

    public int Value { get; }

    public BinarySearchTree Left { get; private set; }

    public BinarySearchTree Right { get; private set; }

    public IEnumerator<int> GetEnumerator()
    {
        if(Left != null)
        {
            var leftEnumerator = Left.GetEnumerator();
            while(leftEnumerator.MoveNext())
                yield return leftEnumerator.Current;
        }

        yield return Value;

        if(Right != null)
        {
            var rightEnumerator = Right.GetEnumerator();
            while(rightEnumerator.MoveNext())
                yield return rightEnumerator.Current;
        }
    }

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
}

File: go/binary-search-tree/binary_search_tree.go
package binarysearchtree

type SearchTreeData struct {
	left  *SearchTreeData
	data  int
	right *SearchTreeData
}

func Bst(n int) *SearchTreeData {
	return &SearchTreeData{nil, n, nil}
}

func (s *SearchTreeData) Insert(v int) {
	if v <= s.data {
		if s.left == nil {
			s.left = Bst(v)
		} else {
			s.left.Insert(v)
		}
	} else {
		if s.right == nil {
			s.right = Bst(v)
		} else {
			s.right.Insert(v)
		}
	}
}

func (s *SearchTreeData) MapString(m func(int) string) []string {
	r := []string{m(s.data)}
	if s.left != nil {
		r = append(s.left.MapString(m), r...)
	}
	if s.right != nil {
		r = append(r, s.right.MapString(m)...)
	}
	return r
}

func (s *SearchTreeData) MapInt(m func(int) int) []int {
	r := []int{m(s.data)}
	if s.left != nil {
		r = append(s.left.MapInt(m), r...)
	}
	if s.right != nil {
		r = append(r, s.right.MapInt(m)...)
	}
	return r
}
